### PR TITLE
[ENH] in `ARCH`, fix `str` coercion of `pd.Series` name

### LINKED ISSUE #5349: [BUG] `test_predict_proba` is expecting column name of `pred_dist` to be `0` always.
**Describe the bug**

All the tests in `check_estimator` are expecting column names to be preserved, while a single test `check_predict_proba` function, invoked inside `test_predict_proba`, in `sktime.forecasting.tests.test_all_forecasters.py` is expecting column name of `pred_dist` i.e., the output of `predict_proba`, to be `0` always.

**To Reproduce**

location: [here](https://github.com/sktime/sktime/blob/12d2770c213382f1ee99bb05a03f19e7ab4cb5e6/sktime/forecasting/tests/test_all_forecasters.py#L529).
```python
# existing code
if isinstance(y_train, pd.Series):
    assert (pred_cols == pd.Index([0])).all()
else:
    assert (pred_cols == y_train.columns).all()

# expected code
if isinstance(y_train, pd.Series):
    assert (pred_cols == y_train.name).all()
else:
    assert (pred_cols == y_train.columns).all()
```

**Expected behavior**

The function mentioned above has to be modified such that, it checks for column names of `y_train`, `pred_dist`, like rest of the tests.

**Additional context**

Any changes in this test will affect existing univariate probabilistic forecasters such as `NaiveForecaster`. Therefore, fixing this issue also requires necessary changes in all the univariate probabilistic forecasters.

### PR DESCRIPTION
**Pull request by Vasudeva-bit**
#### Reference Issues/PRs

Related to #5349 (issue), #5384 (PR).

#### What does this implement/fix? Explain your changes.

> After debugging for a while,
> 
> 1. Firstly, the function `_check_predict_proba` is only testing for `y_train` generated from `_make_data(n_columns=n_columns)`. If the forecaster is univariate, then the `y_train` generated would always have column name `None` of type `str`. The `_check_predict_proba` checks whether the `_predict_proba` is changing `None` to `0` internally. Since every forecaster's `_predict_proba` doing it, all forecasters passed.
> 2. Actually `ARCH` should fail this test, because `_predict_proba` function is deleted in this PR. But, the `type(y_train.name)` is `str` (`'None'`), not `None`. Therefore, `None` case is skipped and checks whether `(pred_cols == y_train.name).all()` which is `None` of `str` equals `None` of `str` so, `ARCH` also passed.
> 
> I think it's my bad, there is no bug in `_check_predict_proba` as it is not designed to check whether column names are preserved. Changes should only be in `ARCH`'s `_predict_proba` like below:
> 
> 1. If the column name is `None` of type `None`/`str`, do `pred_dist.name = pd.Index([0])`.
> 2. Otherwise, just return output of `Super()._predict_proba()` as it is, i.e., column names are preserved by default.

I think there was some ambiguity in my explanation for #5349 (issue) in #5384 (PR), but what I meant was exactly the changes in this PR.

#### What should a reviewer concentrate their feedback on?

Kindly notice that the current fix (#5384) would not check for `None` of `str` type, it only checks for `None` of `None` type. Due to which, user might see `pred_dist` with column name `None`.  Here, `predict_proba` (in `_uarch.py`) is needed to set the column name to `0` if `y_train.name` is either `None` of `None` type or `None` of `str` type.
@fkiraly

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Hm, if `y_train.name` is the string `"None"`, then imo the `y_pred.columns` should also be string `"None"` rather than `0`. This is to ensure case consistency for string, otherwise we would be introducing a "special string" in the contract and making it much more complex.

Though, where in the test cases is a `name` of (string) `"None"` generated? Perhaps that should simply be `None` (none-type, not string)

**Comment by Vasudeva-bit** [comment]
Ohkay!
[Here](https://github.com/sktime/sktime/blob/main/sktime/forecasting/tests/test_all_forecasters.py#L556), `_make_series` is generating `y_train` data with column name `None` of `str` type. I have cross-checked it again.

I think, `predict_proba` (in `_uarch.py`) is needed to set the column name to `0` if `y_train.name` is `None`. Shall I change it accordingly?
Otherwise, we can close this PR.

**Comment by fkiraly** [comment]
>I think, predict_proba (in _uarch.py) is needed to set the column name to 0 if y_train.name is None. Shall I change it accordingly?

I would say:

* let's change `predict_proba`
* let's ensure we test somewhere with `y_train.name = None` (not str) as well as with a `y_train.name` being some string.

**Comment by Vasudeva-bit** [comment]
> * let's ensure we test somewhere with `y_train.name = None` (not str) as well as with a `y_train.name` being some string.

Do you expect these changes only to `ARCH` or to all the forecasters?


If changes are only to `ARCH`, is this check enough?
```
if pred_dist.columns is None:
            pred_dist.columns = pd.Index([0])
```

**Comment by fkiraly** [comment]
> If changes are only to ARCH, is this check enough?

I think that wouldn't work, we would have to check `y_train`.

I would suggest, let's see first how many forecasters would fail the correct check - then decide whether we need changes to the base class, or just to individual forecasters.

File: sktime/forecasting/tests/test_all_forecasters.py [hunk 1]
```diff
@@ -554,10 +554,16 @@ def test_predict_proba(self, estimator_instance, n_columns, fh_int_oos):
...
 
         if estimator_instance.get_tag("capability:pred_int"):
             try:
+                # y_train.name is "None" (str)
+                estimator_instance.fit(y_train, fh=fh_int_oos)
+                pred_dist = estimator_instance.predict_proba()
+                self._check_predict_proba(pred_dist, y_train, fh_int_oos)
+                # y_train.name is None (None)
+                estimator_instance.fit(y_train, fh=fh_int_oos)
+                y_train.name = None
```
Comment (fkiraly) [inline sktime/forecasting/tests/test_all_forecasters.py:566]: are these two lines not in the wrong order? I.e., set the `name` and then pass it to `fit`?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks! Looks ok, except two lines seem to be in wrong order.

File: sktime/forecasting/tests/test_all_forecasters.py (hunk 1 above)
Comment (Vasudeva-bit) [reply sktime/forecasting/tests/test_all_forecasters.py:566]: Will modify, test and push the changes.

Comment (Vasudeva-bit) [reply sktime/forecasting/tests/test_all_forecasters.py:566]: The data generated by `_make_data(n_columns=n_columns)` has name `None` of type `None` initially (`y_train.name`).
For some reason, the `fit` method is causing this:
```
Before fit:  None <class 'NoneType'>
After fit:  None <class 'str'>
```
Code to reproduce.
```
# Replace line 565 with following three lines in test_all_forecasters.py of this PR.

print("Before fit: ", y_train.name, type(y_train.name))
estimator_instance.fit(y_train, fh=fh_int_oos)
print("After fit: ", y_train.name, type(y_train.name))
```

Comment (fkiraly) [reply sktime/forecasting/tests/test_all_forecasters.py:566]: of which estimator, `ARCH`?
Do you know if this happens with all estimators, or only some?

What's strange is that we have tests for no side effects, these should catch `y_train` being mutated.

More precisely, for all scenarios (data) in `utils._testing.scenarios_forecasting` (that have the "enabled" tag), we test that none of the methods mutate the data. So it might be interesting to understand why it's not being caught.

Comment (Vasudeva-bit) [reply sktime/forecasting/tests/test_all_forecasters.py:566]: Ohkay, I will take a closer look at it.

**Comment by fkiraly** [comment]
@Vasudeva-bit, are you still working on this?

**Comment by Vasudeva-bit** [comment]
Yes. Sorry for the delay, I am giving semester exams. I will be free very soon. I hope to complete all my pending tasks within a week.

File: sktime/forecasting/tests/test_all_forecasters.py (hunk 1 above)
Comment (Vasudeva-bit) [reply sktime/forecasting/tests/test_all_forecasters.py:566]: This isn't the case with any estimator other than `ARCH`. This effect was caused by `arch` (`arch_model`). I fixed it by passing deep copies of `X`, `y` to 'arch_model' rather than just `X`, `y`. Kindly review the code.

File: sktime/forecasting/arch/_uarch.py [hunk 2]
```diff
@@ -344,7 +344,7 @@ def _predict(self, fh, X=None):
         y_pred = pd.Series(
             ArchResultObject.mean.values[-1],
             index=full_range,
-            name=str(self._y.name),
```
Comment (fkiraly) [inline sktime/forecasting/arch/_uarch.py:347]: ah, that was the culprit!

File: sktime/forecasting/arch/_uarch.py [hunk 3]
```diff
@@ -252,8 +252,8 @@ def _fit(self, y, X=None, fh=None):
             self._horizon = fh
 
         self._forecaster = _ARCH(
-            y=y,
-            x=X,
+            y=y.copy(deep=True),
+            x=X.copy(deep=True) if X is not None else X,
```
Comment (fkiraly) [inline sktime/forecasting/arch/_uarch.py:256]: why do we need to make copies?
Does `_ARCH` have side effects? If yes, I would report this as a bug in the arch package.

File: sktime/forecasting/arch/_uarch.py [hunk 4]
```diff
@@ -398,19 +398,19 @@ def _predict_interval(self, fh, X, coverage):
             upper_int = mean_forecast + (z_critical * std_err)
             lower_df = pd.DataFrame(
                 lower_int,
-                columns=[y_col_name + " " + str(alpha) + " " + "lower"],
+                columns=[str(y_col_name) + " " + str(alpha) + " " + "lower"],
```
Comment (fkiraly) [inline sktime/forecasting/arch/_uarch.py:401]: we should not convert to str - convention is, names should be identical to column names in the input `y`

File: sktime/forecasting/arch/_uarch.py [hunk 5]
```diff
@@ -464,6 +464,29 @@ def _predict_var(self, fh=None, X=None, cov=False):
...
+            Exogeneous time series for the forecast
+        marginal : bool, optional (default=True)
+            whether returned distribution is marginal by time index
+
+        Returns
+        -------
+        pred_dist : sktime BaseDistribution
+            predictive distribution
+        """
+        pred_dist = super()._predict_proba(fh=fh, X=X, marginal=marginal)
```
Comment (fkiraly) [inline sktime/forecasting/arch/_uarch.py:485]: why is this needed?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Left some comments - the main one is, we should not convert column names to string if we can avoid.

File: sktime/forecasting/arch/_uarch.py (hunk 3 above)
Comment (Vasudeva-bit) [reply sktime/forecasting/arch/_uarch.py:256]: image
```
import pandas as pd
from sktime.datasets import load_airline
y = load_airline()
y.name = None
print("Before fit: ", y.name, type(y.name))
from arch import arch_model
model = arch_model(y, rescale=True)
print("After fit: ", y.name, type(y.name))
```
Yes, bug need to be reported.

File: sktime/forecasting/arch/_uarch.py (hunk 4 above)
Comment (Vasudeva-bit) [reply sktime/forecasting/arch/_uarch.py]: Kindly review the changes.

File: sktime/forecasting/arch/_uarch.py (hunk 5 above)
Comment (fkiraly) [reply sktime/forecasting/arch/_uarch.py:489]: but why does the `super` not automatically take care of it?

File: sktime/forecasting/arch/_uarch.py (hunk 3 above)
Comment (fkiraly) [reply sktime/forecasting/arch/_uarch.py:256]: is this only `ARCH`, or other estimators in general?

If only `ARCH`, we could overwrite the `name` at the end of `_fit`, instead of creating a full copy?

Comment (Vasudeva-bit) [reply sktime/forecasting/arch/_uarch.py:256]: Other estimators are fine, it's only `ARCH`. Let's overwrite the `name` at the end of `_fit` as you said.

File: sktime/forecasting/arch/_uarch.py (hunk 5 above)
Comment (Vasudeva-bit) [reply sktime/forecasting/arch/_uarch.py]: `super` takes care of it. I should have looked into that closely.

**Review by fkiraly — APPROVED** [review]
Thanks for the explanation.
